cocotb/share/lib/verilator/verilator.cpp

```diff
@@ -56,44 +75,68 @@ int main(int argc, char** argv) {
 #endif
 #endif
 
+    bool again;
     while (!Verilated::gotFinish()) {
-        bool again = true;
+        // Call registered timed callbacks (e.g. clock timer)
+        // These are called at the beginning of the time step
+        // before the iterative regions (IEEE 1800-2012 4.4.1)
+        VerilatedVpi::callTimedCbs();
+
+        // Call Value Change callbacks triggered by Timer callbacks
+        // These can modify signal values
+        settle_value_callbacks();
 
         // We must evaluate whole design until we process all 'events'
+        again = true;
         while (again) {
             // Evaluate design
             top->eval();
 
-            // Call Value Change callbacks as eval()
-            // can modify signals values
-            VerilatedVpi::callValueCbs();
+            // Call Value Change callbacks triggered by eval()
+            // These can modify signal values
+            again = settle_value_callbacks();
```

Comment:
If there are every any value change callbacks added in any of the following statements, this will evaluate to true and you will enter an infinite loop. I used the original logic where only the readwrite callbacks were considered for running again, and the examples worked flawlessly.

The loop 
1. Evaluates the model to allow it to react to changes by the user
2. Allow the user to react to value changes by the model and cache up writes of their own.
3. ReadWrite occurs and cached writes are flushed
4. Allow the user to react to value changes by the user and potentially cache more writes
5. Restart from 1.

The process stops when everyone is done changing. There are two exit conditions: no values changed in the model (we don't know that since `eval` doesn't return anything), or the user is done making changes (we use the fact there were no `ReadWrite` callbacks for that, since cached writes are flushed in that stage). Whether the user needed to react to changes or not doesn't matter, what matters is if they changed anything. Because of `setimmediate` this is always possible, but we have to draw a line in the sand somewhere.

Reply:
There's a possibility of an infinite loop if something keeps toggling/glitching back and forth, but I think that scenario can occur in most/all simulators.

When a `cbValueChange` callback is registered, verilator stores the value at that time as the previous value to compare against for a value change. After `callValueCbs` loops through the registered callbacks, those previous values are updated with the changed value. So even new value change callbacks registered during a previous value change callback on the same object should not be called unless the value changes again, between when it's registered and `callValueCbs` is called subsequently.

`axi_lite_slave` isn't getting stuck in the loop in a single time step, it keeps going forever, waiting for the end of the transaction. For some reason the `AXIML_BVALID` signal is not high during the `ReadOnly` callback when `AXI4Master.write` is looking for it. I don't know the root cause of that though.